pkg/admission/prep_downscale.go

```diff
@@ -155,27 +164,54 @@ func prepareDownscale(ctx context.Context, logger log.Logger, ar v1.AdmissionRev
 			)
 		}
 
-		foundSts, err := findDownscalesDoneMinTimeAgo(stsList, ar.Request.Name)
-		if err != nil {
-			level.Warn(logger).Log("msg", "downscale not allowed due to error while parsing downscale annotations", "err", err)
-			return deny(
-				"downscale of %s/%s in %s from %d to %d replicas is not allowed because parsing downscale annotations failed.",
-				ar.Request.Resource.Resource, ar.Request.Name, ar.Request.Namespace, *oldReplicas, *newReplicas,
-			)
-		}
-		if foundSts != nil {
-			msg := fmt.Sprintf("downscale of %s/%s in %s from %d to %d replicas is not allowed because statefulset %v was downscaled at %v and is labelled to wait %s between zone downscales",
-				ar.Request.Resource.Resource, ar.Request.Name, ar.Request.Namespace, *oldReplicas, *newReplicas, foundSts.name, foundSts.lastDownscaleTime, foundSts.waitTime)
-			level.Warn(logger).Log("msg", msg, "err", err)
-			return deny(msg)
-		}
+		// If using zoneTracker instead of downscale annotations, check the zone file for recent downscaling.
+		// Otherwise, check the downscale annotations.
+		if zt != nil {
+			if err := zt.loadZones(ctx, stsList); err != nil {
+				level.Warn(logger).Log("msg", "downscale not allowed due to error while loading zones", "err", err)
+				return deny(
+					"downscale of %s/%s in %s from %d to %d replicas is not allowed because loading zones failed.",
+					ar.Request.Resource.Resource, ar.Request.Name, ar.Request.Namespace, *oldReplicas, *newReplicas,
+				)
+			}
+			// Check if the zone has been downscaled recently.
+			foundSts, err := zt.findDownscalesDoneMinTimeAgo(stsList, ar.Request.Name)
+			if err != nil {
+				level.Warn(logger).Log("msg", "downscale not allowed due to error while parsing downscale timestamps from the zone file", "err", err)
+				return deny(
+					"downscale of %s/%s in %s from %d to %d replicas is not allowed because parsing parsing downscale timestamps from the zone file failed.",
+					ar.Request.Resource.Resource, ar.Request.Name, ar.Request.Namespace, *oldReplicas, *newReplicas,
+				)
+			}
+			if foundSts != nil {
+				msg := fmt.Sprintf("downscale of %s/%s in %s from %d to %d replicas is not allowed because statefulset %v was downscaled at %v and is labelled to wait %s between zone downscales",
+					ar.Request.Resource.Resource, ar.Request.Name, ar.Request.Namespace, *oldReplicas, *newReplicas, foundSts.name, foundSts.lastDownscaleTime, foundSts.waitTime)
+				level.Warn(logger).Log("msg", msg, "err", err)
+				return deny(msg)
+			}
+		} else {
+			foundSts, err := findDownscalesDoneMinTimeAgo(stsList, ar.Request.Name)
```

Comment:
Do we still need the old way? If the initial load of the zones adds the waiting time to the initial times won't that be enough to migrate?